internal/services/storage/storage_account_resource.go

```diff
@@ -1900,7 +1900,7 @@ func resourceStorageAccountRead(d *pluginsdk.ResourceData, meta interface{}) err
 		if e, ok := err.(azautorest.DetailedError); ok {
 			if status, ok := e.StatusCode.(int); ok {
 				hasWriteLock = status == http.StatusConflict
-				doesntHavePermissions = status == http.StatusUnauthorized
+				doesntHavePermissions = (status == http.StatusUnauthorized || status == http.StatusForbidden)
```

Comment:
@magodo it'd be worth spinning one up to check, but I have a feeling that a 403 is a valid error here for the Resource (but wouldn't be valid for the Data Source, where we allow for less permissive permissions)?

From memory a 401 indicates differently that access keys are disabled on the storage account too?

Reply:
My local test shows that if `shared_access_key_enabled = false`, the `listKeys` call still succeeds, while the follow up calls to service like queue with shared key auth will fail with 403, complaining "Key based authentication is not permitted on this storage account.". Using `storage_use_azuread = true` in the provider block can move on.

Meanwhile, I've created a new app which is only assigned following roles:

- Microsoft.Resources/subscriptions/resourceGroups/read
- Microsoft.Storage/storageAccounts/read

Using this SP to read the storage account via the provider returns 403:

> │ Error: listing Keys for Storage Account: (Name "unlikely24exst2acctv9ipk" / Resource Group "acctestRG-storage-220816112105435513"): storage.AccountsClient#ListKeys: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'xxxx' with object id 'xxxx' does not have authorization to perform action 'Microsoft.Storage/storageAccounts/listKeys/action' over scope '/subscriptions/xxxx/resourceGroups/acctestRG-storage-220816112105435513/providers/Microsoft.Storage/storageAccounts/unlikely24exst2acctv9ipk' or the scope is invalid. If access was recently granted, please refresh your credentials."

In the end, I didn't figure out when do we get 401.

Reply:
sounds reasonable to me, let's leave the 401 thing in for now 👍